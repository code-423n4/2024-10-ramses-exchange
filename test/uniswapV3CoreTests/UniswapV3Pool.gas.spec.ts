import { ethers } from "hardhat";
import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
import { Wallet } from "ethers";
import { MockTimeRamsesV3Pool } from "../../typechain-types";
import { expect } from "./shared/expect";

import { poolFixture } from "./shared/fixtures";
import snapshotGasCost from "./shared/snapshotGasCost";

import {
    expandTo18Decimals,
    FeeAmount,
    getMinTick,
    encodePriceSqrt,
    TICK_SPACINGS,
    createPoolFunctions,
    SwapFunction,
    MintFunction,
    getMaxTick,
    MaxUint128,
    SwapToPriceFunction,
    MAX_SQRT_RATIO,
    MIN_SQRT_RATIO,
} from "./shared/utilities";

describe("RamsesV3Pool gas tests", () => {
    let wallet: Wallet, other: Wallet;

    before("create fixture loader", async () => {
        [wallet, other] = await (ethers as any).getSigners();
    });

    for (const feeProtocol of [0, 6]) {
        describe(feeProtocol > 0 ? "fee is on" : "fee is off", () => {
            const startingPrice = encodePriceSqrt(100001n, 100000n).toString();
            const startingTick = 0;
            const feeAmount = FeeAmount.MEDIUM;
            const tickSpacing = TICK_SPACINGS[feeAmount];
            const minTick = getMinTick(tickSpacing);
            const maxTick = getMaxTick(tickSpacing);

            const gasTestFixture = async () => {
                const fix = await poolFixture();

                const pool = await fix.createPool(tickSpacing, startingPrice);

                const {
                    swapExact0For1,
                    swapToHigherPrice,
                    mint,
                    swapToLowerPrice,
                } = await createPoolFunctions({
                    swapTarget: fix.swapTargetCallee,
                    token0: fix.token0,
                    token1: fix.token1,
                    pool,
                });

                await fix.factory.setFeeCollector(wallet.address);
                await fix.factory.setFeeProtocol(BigInt(feeProtocol));
                await pool.setFeeProtocol();
                await pool.increaseObservationCardinalityNext(4);
                await pool.advanceTime(1);
                await mint(
                    wallet.address,
                    0n,
                    minTick,
                    maxTick,
                    expandTo18Decimals(2n),
                );

                await swapExact0For1(expandTo18Decimals(1n), wallet.address);
                await pool.advanceTime(1);
                await swapToHigherPrice(BigInt(startingPrice), wallet.address);
                await pool.advanceTime(1);
                expect((await pool.slot0()).tick).to.eq(startingTick);
                expect((await pool.slot0()).sqrtPriceX96).to.eq(startingPrice);

                return {
                    pool,
                    swapExact0For1,
                    mint,
                    swapToHigherPrice,
                    swapToLowerPrice,
                };
            };

            let swapExact0For1: SwapFunction;
            let swapToHigherPrice: SwapToPriceFunction;
            let swapToLowerPrice: SwapToPriceFunction;
            let pool: MockTimeRamsesV3Pool;
            let mint: MintFunction;

            beforeEach("load the fixture", async () => {
                ({
                    swapExact0For1,
                    pool,
                    mint,
                    swapToHigherPrice,
                    swapToLowerPrice,
                } = await loadFixture(gasTestFixture));
            });

            describe("#swapExact0For1", () => {
                it("first swap in block with no tick movement", async () => {
                    await snapshotGasCost(
                        swapExact0For1(2000n, wallet.address),
                    );
                    expect((await pool.slot0()).sqrtPriceX96).to.not.eq(
                        startingPrice,
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick);
                });

                it("first swap in block moves tick, no initialized crossings", async () => {
                    await snapshotGasCost(
                        swapExact0For1(
                            expandTo18Decimals(1n) / 10000n,
                            wallet.address,
                        ),
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick - 1);
                });

                it("second swap in block with no tick movement", async () => {
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 10000n,
                        wallet.address,
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick - 1);
                    await snapshotGasCost(
                        swapExact0For1(2000n, wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick - 1);
                });

                it("second swap in block moves tick, no initialized crossings", async () => {
                    await swapExact0For1(1000n, wallet.address);
                    expect((await pool.slot0()).tick).to.eq(startingTick);
                    await snapshotGasCost(
                        swapExact0For1(
                            expandTo18Decimals(1n) / 10000n,
                            wallet.address,
                        ),
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick - 1);
                });

                it("first swap in block, large swap, no initialized crossings", async () => {
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(10n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.eq(-35787);
                });

                it("first swap in block, large swap crossing several initialized ticks", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 3 * tickSpacing,
                        startingTick - tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 4 * tickSpacing,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    expect((await pool.slot0()).tick).to.eq(startingTick);
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        startingTick - 4 * tickSpacing,
                    ); // we crossed the last tick
                });

                it("first swap in block, large swap crossing a single initialized tick", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        minTick,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        startingTick - 2 * tickSpacing,
                    ); // we crossed the last tick
                });

                it("second swap in block, large swap crossing several initialized ticks", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 3 * tickSpacing,
                        startingTick - tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 4 * tickSpacing,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 10000n,
                        wallet.address,
                    );
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        startingTick - 4 * tickSpacing,
                    );
                });

                it("second swap in block, large swap crossing a single initialized tick", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        minTick,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 10000n,
                        wallet.address,
                    );
                    expect((await pool.slot0()).tick).to.be.gt(
                        startingTick - 2 * tickSpacing,
                    ); // we didn't cross the initialized tick
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        startingTick - 2 * tickSpacing,
                    ); // we crossed the last tick
                });

                it("large swap crossing several initialized ticks after some time passes", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 3 * tickSpacing,
                        startingTick - tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 4 * tickSpacing,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(2n, wallet.address);
                    await pool.advanceTime(1);
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        startingTick - 4 * tickSpacing,
                    );
                });

                it("large swap crossing several initialized ticks second time after some time passes", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 3 * tickSpacing,
                        startingTick - tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await mint(
                        wallet.address,
                        0n,
                        startingTick - 4 * tickSpacing,
                        startingTick - 2 * tickSpacing,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(
                        expandTo18Decimals(1n),
                        wallet.address,
                    );
                    await swapToHigherPrice(
                        BigInt(startingPrice),
                        wallet.address,
                    );
                    await pool.advanceTime(1);
                    await snapshotGasCost(
                        swapExact0For1(expandTo18Decimals(1n), wallet.address),
                    );
                    expect((await pool.slot0()).tick).to.be.lt(
                        tickSpacing * -4,
                    );
                });
            });

            describe("#mint", () => {
                for (const { description, tickLower, tickUpper } of [
                    {
                        description: "around current price",
                        tickLower: startingTick - tickSpacing,
                        tickUpper: startingTick + tickSpacing,
                    },
                    {
                        description: "below current price",
                        tickLower: startingTick - 2 * tickSpacing,
                        tickUpper: startingTick - tickSpacing,
                    },
                    {
                        description: "above current price",
                        tickLower: startingTick + tickSpacing,
                        tickUpper: startingTick + 2 * tickSpacing,
                    },
                ]) {
                    describe(description, () => {
                        it("new position mint first in range", async () => {
                            await snapshotGasCost(
                                mint(
                                    wallet.address,
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                        it("add to position existing", async () => {
                            await mint(
                                wallet.address,
                                0n,
                                tickLower,
                                tickUpper,
                                expandTo18Decimals(1n),
                            );
                            await snapshotGasCost(
                                mint(
                                    wallet.address,
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                        it("second position in same range", async () => {
                            await mint(
                                wallet.address,
                                1n,
                                tickLower,
                                tickUpper,
                                expandTo18Decimals(1n),
                            );
                            await snapshotGasCost(
                                mint(
                                    other.address,
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                        it("add to position after some time passes", async () => {
                            await mint(
                                wallet.address,
                                0n,
                                tickLower,
                                tickUpper,
                                expandTo18Decimals(1n),
                            );
                            await pool.advanceTime(1);
                            await snapshotGasCost(
                                mint(
                                    wallet.address,
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                    });
                }
            });

            describe("#burn", () => {
                for (const { description, tickLower, tickUpper } of [
                    {
                        description: "around current price",
                        tickLower: startingTick - tickSpacing,
                        tickUpper: startingTick + tickSpacing,
                    },
                    {
                        description: "below current price",
                        tickLower: startingTick - 2 * tickSpacing,
                        tickUpper: startingTick - tickSpacing,
                    },
                    {
                        description: "above current price",
                        tickLower: startingTick + tickSpacing,
                        tickUpper: startingTick + 2 * tickSpacing,
                    },
                ]) {
                    describe(description, () => {
                        const liquidityAmount = expandTo18Decimals(1n);
                        beforeEach("mint a position", async () => {
                            await mint(
                                wallet.address,
                                0n,
                                tickLower,
                                tickUpper,
                                liquidityAmount,
                            );
                        });

                        it("burn when only position using ticks", async () => {
                            await snapshotGasCost(
                                pool.burn(
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                        it("partial position burn", async () => {
                            await snapshotGasCost(
                                pool.burn(
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n) / 2n,
                                ),
                            );
                        });
                        it("entire position burn but other positions are using the ticks", async () => {
                            await mint(
                                other.address,
                                0n,
                                tickLower,
                                tickUpper,
                                expandTo18Decimals(1n),
                            );
                            await snapshotGasCost(
                                pool.burn(
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                        it("burn entire position after some time passes", async () => {
                            await pool.advanceTime(1);
                            await snapshotGasCost(
                                pool.burn(
                                    0n,
                                    tickLower,
                                    tickUpper,
                                    expandTo18Decimals(1n),
                                ),
                            );
                        });
                    });
                }
            });

            describe("#poke", () => {
                const tickLower = startingTick - tickSpacing;
                const tickUpper = startingTick + tickSpacing;

                it("best case", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        tickLower,
                        tickUpper,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 100n,
                        wallet.address,
                    );
                    await pool.burn(0n, tickLower, tickUpper, 0);
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 100n,
                        wallet.address,
                    );
                    await snapshotGasCost(
                        pool.burn(0n, tickLower, tickUpper, 0),
                    );
                });
            });

            describe("#collect", () => {
                const tickLower = startingTick - tickSpacing;
                const tickUpper = startingTick + tickSpacing;

                it("close to worst case", async () => {
                    await mint(
                        wallet.address,
                        0n,
                        tickLower,
                        tickUpper,
                        expandTo18Decimals(1n),
                    );
                    await swapExact0For1(
                        expandTo18Decimals(1n) / 100n,
                        wallet.address,
                    );
                    await pool.burn(0n, tickLower, tickUpper, 0); // poke to accumulate fees
                    await snapshotGasCost(
                        pool.collect(
                            wallet.address,
                            0n,
                            tickLower,
                            tickUpper,
                            MaxUint128,
                            MaxUint128,
                        ),
                    );
                });
            });

            describe("#increaseObservationCardinalityNext", () => {
                it("grow by 1 slot", async () => {
                    await snapshotGasCost(
                        pool.increaseObservationCardinalityNext(5),
                    );
                });
                it("no op", async () => {
                    await snapshotGasCost(
                        pool.increaseObservationCardinalityNext(3),
                    );
                });
            });

            describe("#snapshotCumulativesInside", () => {
                it("tick inside", async () => {
                    await snapshotGasCost(
                        pool.snapshotCumulativesInside.estimateGas(
                            minTick,
                            maxTick,
                        ),
                    );
                });
                it("tick above", async () => {
                    await swapToHigherPrice(
                        MAX_SQRT_RATIO - 1n,
                        wallet.address,
                    );
                    await snapshotGasCost(
                        pool.snapshotCumulativesInside.estimateGas(
                            minTick,
                            maxTick,
                        ),
                    );
                });
                it("tick below", async () => {
                    await swapToLowerPrice(MIN_SQRT_RATIO + 1n, wallet.address);
                    await snapshotGasCost(
                        pool.snapshotCumulativesInside.estimateGas(
                            minTick,
                            maxTick,
                        ),
                    );
                });
            });
        });
    }
});
