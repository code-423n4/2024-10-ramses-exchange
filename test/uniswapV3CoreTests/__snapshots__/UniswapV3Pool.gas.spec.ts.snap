// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RamsesV3Pool gas tests fee is off #burn above current price burn entire position after some time passes 1`] = `120044`;

exports[`RamsesV3Pool gas tests fee is off #burn above current price burn when only position using ticks 1`] = `120044`;

exports[`RamsesV3Pool gas tests fee is off #burn above current price entire position burn but other positions are using the ticks 1`] = `120422`;

exports[`RamsesV3Pool gas tests fee is off #burn above current price partial position burn 1`] = `130022`;

exports[`RamsesV3Pool gas tests fee is off #burn around current price burn entire position after some time passes 1`] = `156627`;

exports[`RamsesV3Pool gas tests fee is off #burn around current price burn when only position using ticks 1`] = `134910`;

exports[`RamsesV3Pool gas tests fee is off #burn around current price entire position burn but other positions are using the ticks 1`] = `128605`;

exports[`RamsesV3Pool gas tests fee is off #burn around current price partial position burn 1`] = `138205`;

exports[`RamsesV3Pool gas tests fee is off #burn below current price burn entire position after some time passes 1`] = `131449`;

exports[`RamsesV3Pool gas tests fee is off #burn below current price burn when only position using ticks 1`] = `131449`;

exports[`RamsesV3Pool gas tests fee is off #burn below current price entire position burn but other positions are using the ticks 1`] = `123479`;

exports[`RamsesV3Pool gas tests fee is off #burn below current price partial position burn 1`] = `133079`;

exports[`RamsesV3Pool gas tests fee is off #collect close to worst case 1`] = `45892`;

exports[`RamsesV3Pool gas tests fee is off #increaseObservationCardinalityNext grow by 1 slot 1`] = `50681`;

exports[`RamsesV3Pool gas tests fee is off #increaseObservationCardinalityNext no op 1`] = `24310`;

exports[`RamsesV3Pool gas tests fee is off #mint above current price add to position after some time passes 1`] = `138917`;

exports[`RamsesV3Pool gas tests fee is off #mint above current price add to position existing 1`] = `138917`;

exports[`RamsesV3Pool gas tests fee is off #mint above current price new position mint first in range 1`] = `335294`;

exports[`RamsesV3Pool gas tests fee is off #mint above current price second position in same range 1`] = `233791`;

exports[`RamsesV3Pool gas tests fee is off #mint around current price add to position after some time passes 1`] = `201147`;

exports[`RamsesV3Pool gas tests fee is off #mint around current price add to position existing 1`] = `171149`;

exports[`RamsesV3Pool gas tests fee is off #mint around current price new position mint first in range 1`] = `440013`;

exports[`RamsesV3Pool gas tests fee is off #mint around current price second position in same range 1`] = `266023`;

exports[`RamsesV3Pool gas tests fee is off #mint below current price add to position after some time passes 1`] = `142455`;

exports[`RamsesV3Pool gas tests fee is off #mint below current price add to position existing 1`] = `142455`;

exports[`RamsesV3Pool gas tests fee is off #mint below current price new position mint first in range 1`] = `419903`;

exports[`RamsesV3Pool gas tests fee is off #mint below current price second position in same range 1`] = `237329`;

exports[`RamsesV3Pool gas tests fee is off #poke best case 1`] = `82257`;

exports[`RamsesV3Pool gas tests fee is off #snapshotCumulativesInside tick above 1`] = `35771`;

exports[`RamsesV3Pool gas tests fee is off #snapshotCumulativesInside tick below 1`] = `35725`;

exports[`RamsesV3Pool gas tests fee is off #snapshotCumulativesInside tick inside 1`] = `41413`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `120932`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 first swap in block with no tick movement 1`] = `105772`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `168199`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `287173`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `134434`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `287173`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `238033`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `120932`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 second swap in block with no tick movement 1`] = `105964`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `154372`;

exports[`RamsesV3Pool gas tests fee is off #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `274197`;

exports[`RamsesV3Pool gas tests fee is on #burn above current price burn entire position after some time passes 1`] = `120044`;

exports[`RamsesV3Pool gas tests fee is on #burn above current price burn when only position using ticks 1`] = `120044`;

exports[`RamsesV3Pool gas tests fee is on #burn above current price entire position burn but other positions are using the ticks 1`] = `120422`;

exports[`RamsesV3Pool gas tests fee is on #burn above current price partial position burn 1`] = `130022`;

exports[`RamsesV3Pool gas tests fee is on #burn around current price burn entire position after some time passes 1`] = `156627`;

exports[`RamsesV3Pool gas tests fee is on #burn around current price burn when only position using ticks 1`] = `134910`;

exports[`RamsesV3Pool gas tests fee is on #burn around current price entire position burn but other positions are using the ticks 1`] = `128605`;

exports[`RamsesV3Pool gas tests fee is on #burn around current price partial position burn 1`] = `138205`;

exports[`RamsesV3Pool gas tests fee is on #burn below current price burn entire position after some time passes 1`] = `131449`;

exports[`RamsesV3Pool gas tests fee is on #burn below current price burn when only position using ticks 1`] = `131449`;

exports[`RamsesV3Pool gas tests fee is on #burn below current price entire position burn but other positions are using the ticks 1`] = `123479`;

exports[`RamsesV3Pool gas tests fee is on #burn below current price partial position burn 1`] = `133079`;

exports[`RamsesV3Pool gas tests fee is on #collect close to worst case 1`] = `45892`;

exports[`RamsesV3Pool gas tests fee is on #increaseObservationCardinalityNext grow by 1 slot 1`] = `50681`;

exports[`RamsesV3Pool gas tests fee is on #increaseObservationCardinalityNext no op 1`] = `24310`;

exports[`RamsesV3Pool gas tests fee is on #mint above current price add to position after some time passes 1`] = `138917`;

exports[`RamsesV3Pool gas tests fee is on #mint above current price add to position existing 1`] = `138917`;

exports[`RamsesV3Pool gas tests fee is on #mint above current price new position mint first in range 1`] = `335294`;

exports[`RamsesV3Pool gas tests fee is on #mint above current price second position in same range 1`] = `233791`;

exports[`RamsesV3Pool gas tests fee is on #mint around current price add to position after some time passes 1`] = `201147`;

exports[`RamsesV3Pool gas tests fee is on #mint around current price add to position existing 1`] = `171149`;

exports[`RamsesV3Pool gas tests fee is on #mint around current price new position mint first in range 1`] = `440013`;

exports[`RamsesV3Pool gas tests fee is on #mint around current price second position in same range 1`] = `266023`;

exports[`RamsesV3Pool gas tests fee is on #mint below current price add to position after some time passes 1`] = `142455`;

exports[`RamsesV3Pool gas tests fee is on #mint below current price add to position existing 1`] = `142455`;

exports[`RamsesV3Pool gas tests fee is on #mint below current price new position mint first in range 1`] = `419903`;

exports[`RamsesV3Pool gas tests fee is on #mint below current price second position in same range 1`] = `237329`;

exports[`RamsesV3Pool gas tests fee is on #poke best case 1`] = `82257`;

exports[`RamsesV3Pool gas tests fee is on #snapshotCumulativesInside tick above 1`] = `35771`;

exports[`RamsesV3Pool gas tests fee is on #snapshotCumulativesInside tick below 1`] = `35725`;

exports[`RamsesV3Pool gas tests fee is on #snapshotCumulativesInside tick inside 1`] = `41413`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `126494`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 first swap in block with no tick movement 1`] = `105901`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `173890`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `293251`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `140254`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `293251`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `244111`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `126494`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 second swap in block with no tick movement 1`] = `106093`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `159934`;

exports[`RamsesV3Pool gas tests fee is on #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `280146`;
